sql/core/src/main/scala/org/apache/spark/sql/Observation.scala

```diff
@@ -0,0 +1,170 @@
+
+  private val listener: ObservationListener = ObservationListener(this)
+
+  private var sparkSession: Option[SparkSession] = None
+
+  @volatile private var row: Option[Row] = None
+
+  /**
+   * Attaches this observation to the given [[Dataset]] to observe aggregation expressions.
+   *
+   * @param ds dataset
+   * @param expr first aggregation expression
+   * @param exprs more aggregation expressions
+   * @tparam T dataset type
+   * @return observed dataset
+   * @throws IllegalArgumentException If this is a streaming Dataset (ds.isStreaming == true)
+   */
+  def on[T](ds: Dataset[T], expr: Column, exprs: Column*): Dataset[T] = {
+    if (ds.isStreaming) {
+      throw new IllegalArgumentException("Observation does not support streaming Datasets")
+    }
+    register(ds.sparkSession)
+    ds.observe(name, expr, exprs: _*)
+  }
+
+  /**
+   * Waits for the first action on the observed dataset to complete and returns true.
+   * The result is then available through the get method.
+   * This method times out after the given amount of time returning false.
+   *
+   * @param time timeout
+   * @param unit timeout time unit
+   * @return true if action completed within timeout, false otherwise
+   * @throws InterruptedException interrupted while waiting
+   */
+  def waitCompleted(time: Long, unit: TimeUnit): Boolean = waitCompleted(Some(unit.toMillis(time)))
+
+  /**
+   * Get the observed metrics. This waits until the observed dataset finishes its first action.
+   * If you want to wait for the result and provide a timeout, use [[waitCompleted]]. Only the
+   * result of the first action is available. Subsequent actions do not modify the result.
+   *
+   * @return the observed metrics as a [[Row]]
+   * @throws InterruptedException interrupted while waiting
+   */
+  def get: Row = {
+    assert(waitCompleted(None), "waitCompleted without timeout returned false")
+    row.get
+  }
+
+  private def waitCompleted(millis: Option[Long]): Boolean = {
+    synchronized {
+      if (row.isEmpty) {
+        if (millis.isDefined) {
+          this.wait(millis.get)
+        } else {
+          this.wait()
+        }
+      }
+      row.isDefined
+    }
+  }
+
+  private def register(sparkSession: SparkSession): Unit = {
+    // makes this class thread-safe:
+    // only the first thread entering this block can set sparkSession
+    // all other threads will see the exception, because it is only allowed to do this once
+    synchronized {
+      if (this.sparkSession.isDefined) {
+        throw new IllegalStateException("An Observation can be used with a Dataset only once")
+      }
+      this.sparkSession = Some(sparkSession)
+    }
+
+    sparkSession.listenerManager.register(listener)
+  }
+
+  private def unregister(): Unit = {
+    this.sparkSession.foreach(_.listenerManager.unregister(listener))
+  }
+
+  private[spark] def onFinish(qe: QueryExecution): Unit = {
+    synchronized {
+      this.row = qe.observedMetrics.get(name)
```

Comment:
It's possible that two DataFrame actions happen within a short time and we receive the second execution success event before we unregister the listener. It's safer to do
```
if (this.row.isEmpty) {
  this.row = qe.observedMetrics.get(name)
  assert(this.row.isDefined, "No metric provided by QueryExecutionListener")
}
this.notifyAll()
```

Reply:
Good point! I have fixed that.

Can there be the situation where an action running before we add our metric to the dataframe calls into our listener because we register the callback just before the action finishes? Would that cause a situation where our listener gets called but `qe.observedMetrics` does not contain our metric? Should we better be not so picky about ` assert(this.row.isDefined) here?

Reply:
> Should we better be not so picky about assert(this.row.isDefined) here?

Good point! We should unregister the listener only if we see a execution success even with required metrics.

Reply:
For one action, there will be exactly one onSuccess or onFailure call, but not both, right? There won't be calls into `onSuccess` or `onFailure` that refer to actions on other Datasets, that still have our metric in them?

I think we should get the metric and only if our metric is in the `QueryExecution`, we notify and unregister, no matter if `onSuccess` or `onFailure`. This allows us to handle calls for other actions that do not contain our metric. And users can get the metrics on failure as well, which might be useful after all.

If the user wants to re-try the failed action on the Dataset while getting the metrics again, they have to first add a new observation to it. Then we can keep our constraint that `Observation` returns the first seen metrics only, while supporting this use case of getting metrics for a failed and subsequent successful action.